app/controllers/topics_controller.rb

```diff
@@ -607,20 +607,37 @@ def set_notifications
   end
 
   def merge_topic
-    params.require(:destination_topic_id)
+    topic_id = params.require(:topic_id)
+    destination_topic_id = params.require(:destination_topic_id)
+    params.permit(:participants)
+    params.permit(:archetype)
 
-    topic = Topic.find_by(id: params[:topic_id])
+    raise Discourse::InvalidAccess if params[:archetype].present? && params[:archetype] == "private_message" && !guardian.is_staff?
```

Comment:
You shouldn't need to check if archetype is present. Checking `== "private_message"` would work even if archetype is `nil`.